pymc3/distributions/mixture.py

```diff
@@ -160,35 +173,100 @@ def _comp_modes(self):
+                      broadcast_shape=None):
+        if self.is_multidim_comp:
+            samples = self._comp_dists.random(point=point, size=size)
         else:
-            # We must iterate the calls to random manually
-            size = to_tuple(size)
-            _size = int(np.prod(size))
-            try:
-                samples = np.array([self.comp_dists.random(point=point,
-                                                           size=None)
-                                    for _ in range(_size)])
-                samples = np.reshape(samples, size + samples.shape[1:])
-            except AttributeError:
-                samples = np.array([[comp_dist.random(point=point, size=None)
-                                     for _ in range(_size)]
-                                    for comp_dist in self.comp_dists])
-                samples = np.moveaxis(samples, 0, samples.ndim - 1)
-                samples = np.reshape(samples, size + samples[1:])
-
+            if comp_dist_shapes is None:
+                comp_dist_shapes = self._comp_dist_shapes
+            if broadcast_shape is None:
+                broadcast_shape = self._sample_shape
+            samples = []
+            for dist_shape, comp_dist in zip(comp_dist_shapes,
+                                             self.comp_dists):
+                def generator(*args, **kwargs):
+                    # The distribution random methods use the size argument
+                    # differently from scipy.*.rvs, and generate_samples
+                    # follows the latter usage pattern. For this reason we
+                    # decorate (horribly hack) the size kwarg of
+                    # comp_dist.random. We also have to disable pylint W0640
+                    # because comp_dist is changed at each iteration of the
+                    # for loop, and this generator function must be defined
+                    # for each comp_dist.
+                    # pylint: disable=W0640
+                    if len(args) > 2:
+                        args[1] = size
+                    else:
+                        kwargs['size'] = size
+                    return comp_dist.random(*args, **kwargs)
+                sample = generate_samples(
+                    generator=generator,
+                    dist_shape=dist_shape,
+                    broadcast_shape=broadcast_shape,
+                    point=point,
+                    size=size,
+                )
+                samples.append(sample)
+            samples = np.array(
+                broadcast_distribution_samples(samples, size=size)
+            )
+            # In the logp we assume the last axis holds the mixture components
+            # so we move the axis to the last dimension
+            samples = np.moveaxis(samples, 0, -1)
         if samples.shape[-1] == 1:
             return samples[..., 0]
         else:
             return samples
 
+    def infer_comp_dist_shapes(self, point=None):
```

Comment:
Maybe it is better to call this once in `__init__` and save the outputs as properties for the mixture class? What do you think?

Reply:
The problem is that the `comp_dist` shapes may depend on the values passed with the `point` dictionary. It's the case for `sample_posterior_predictive` at least.

Reply:
I see, good point.